dace/runtime/include/dace/cuda/dynmap.cuh

```diff
@@ -22,228 +22,205 @@
-    /**
-     * A map (usually dynamically sized) that can be rescheduled across a 
-     * threadblock
-     **/
-    template<int BLOCK_SIZE, typename index_type = int32_t, bool WARP_INTRINSICS = true>
-    struct DynamicMap
-    {
-        template<const int WARPS_PER_TB> struct warp_np {
-            volatile index_type owner[WARPS_PER_TB];
-            volatile index_type start[WARPS_PER_TB];
-            volatile index_type size[WARPS_PER_TB];
-            volatile index_type src[WARPS_PER_TB];
+#include <cooperative_groups.h>
+
+namespace dace {
+
+    namespace cg = cooperative_groups;
+
+    template<bool DYNAMIC, bool FINE_GRAINED, int BLOCK_SIZE, int WARP_SIZE = 32, typename index_type = int32_t>
+    struct DynamicMap {
+
+        // empty field so the compiler doesn't get confused when empty_type is used instead of fg_type
+        struct empty_type {
+            index_type src[0];  // outer map index
+            index_type data[0]; // inner map index
         };
 
-        struct tb_np {
+        struct tb_type {
             index_type owner;
+            index_type src;
             index_type start;
             index_type size;
-            index_type src;
         };
 
-        struct empty_np {
+        // fine grained needs space to accommodate all thread in the warp having the most possible jobs
+        struct fg_type {
+            index_type src[(BLOCK_SIZE / WARP_SIZE) * WARP_SIZE * WARP_SIZE];  // outer map index
+            index_type data[(BLOCK_SIZE / WARP_SIZE) * WARP_SIZE * WARP_SIZE]; // inner map index
         };
 
+        // depending on configuration of scheduler generate smaller union to save space
+        template <typename tb_type, typename fg_type>
+        union shared_type {
+            tb_type tb; // for tb-level np
+            fg_type fg; // fine-grained schedule
+        };
 
-        template <typename ts_type, typename TTB, typename TWP>
-        union np_shared {
-            // for scans
-            ts_type temp_storage;
 
-            // for tb-level np
-            TTB tb;
+        template<typename Functor>
+        __device__ static void schedule(index_type localStart, index_type localEnd, index_type localSrc, Functor&& work) {
 
-            // for warp-level np
-            TWP warp;
+            typedef union std::conditional<FINE_GRAINED,
+                    shared_type<tb_type, fg_type>,
+                    shared_type<tb_type, empty_type>
+            >::type shared_type;
 
-            // fine-grained schedule (unused)
-            //TFG fg;
-        };
+            // defining other local variables
+            index_type localSize = localEnd - localStart;
 
-        /*
-        * @brief A structure representing a scheduled chunk of work
-        */
-        struct np_local
-        {
-            index_type size; // work size
-            index_type start; // work start
-            index_type src; // work source thread / metadata
-        };
+            if (DYNAMIC) {
```

Comment:
I would change dynamic to some more informative name (block/warp scheduler? load balancing?)

Reply:
Unless you think it's good to keep I would actually remove it entirely as it only served debug purposes and doesn't really make sense in the context of a dynamic scheduler

Reply:
ok